backend/src/apiserver/server/util.go

```diff
@@ -241,3 +244,83 @@ func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.P
+	node := oldNodes[oldNodeID]
+	newNodeName := regex.ReplaceAllString(node.Name, newWf.ObjectMeta.Name)
+	return newWf.NodeID(newNodeName)
+}
+
+// FormulateResubmitWorkflow formulate a new workflow from a previous workflow optionally re-using successful nodes
+func formulateResubmitWorkflow(wf *wfv1.Workflow, randomString util.RandomStringInterface) (*wfv1.Workflow, error) {
+	newWF := wfv1.Workflow{}
+	newWF.TypeMeta = wf.TypeMeta
+
+	// Resubmitted workflow will use generated names
+	if wf.ObjectMeta.GenerateName != "" {
+		newWF.ObjectMeta.GenerateName = wf.ObjectMeta.GenerateName
+	} else {
+		newWF.ObjectMeta.GenerateName = wf.ObjectMeta.Name + "-"
+	}
+	// When resubmitting workflow with memoized nodes, we need to use a predetermined workflow name
+	// in order to formulate the node statuses. Which means we cannot reuse metadata.generateName
+	// The following simulates the behavior of generateName
+	newWF.ObjectMeta.Name = newWF.ObjectMeta.GenerateName + randomString.Get(5)
+
+	// carry over the unmodified spec
+	newWF.Spec = wf.Spec
+
+	// carry over user labels and annotations from previous workflow.
+	// skip any argoproj.io labels except for the controller instanceID label.
+	for key, val := range wf.ObjectMeta.Labels {
+		if strings.HasPrefix(key, workflow.FullName+"/") && key != workflow.FullName+"/controller-instanceid" {
+			continue
+		}
+		if newWF.ObjectMeta.Labels == nil {
+			newWF.ObjectMeta.Labels = make(map[string]string)
+		}
+		newWF.ObjectMeta.Labels[key] = val
+	}
+	for key, val := range wf.ObjectMeta.Annotations {
+		if newWF.ObjectMeta.Annotations == nil {
+			newWF.ObjectMeta.Annotations = make(map[string]string)
+		}
+		newWF.ObjectMeta.Annotations[key] = val
+	}
+
+	// Iterate the previous nodes. If it was successful Pod carry it forward
+	replaceRegexp := regexp.MustCompile("^" + wf.ObjectMeta.Name)
+	newWF.Status.Nodes = make(map[string]wfv1.NodeStatus)
+	for _, node := range wf.Status.Nodes {
+		switch node.Phase {
+		case wfv1.NodeSucceeded, wfv1.NodeSkipped:
+			node.Name = replaceRegexp.ReplaceAllString(node.Name, newWF.ObjectMeta.Name)
+			node.ID = newWF.NodeID(node.Name)
+			node.BoundaryID = convertNodeID(&newWF, replaceRegexp, node.BoundaryID, wf.Status.Nodes)
+			newChildren := make([]string, len(node.Children))
+			for i, childID := range node.Children {
+				newChildren[i] = convertNodeID(&newWF, replaceRegexp, childID, wf.Status.Nodes)
+			}
+			node.Children = newChildren
+			newOutboundNodes := make([]string, len(node.OutboundNodes))
+			for i, outboundID := range node.OutboundNodes {
+				newOutboundNodes[i] = convertNodeID(&newWF, replaceRegexp, outboundID, wf.Status.Nodes)
+			}
+			node.OutboundNodes = newOutboundNodes
+			if node.Type == wfv1.NodeTypePod {
+				node.Phase = wfv1.NodeSkipped
+				node.Type = wfv1.NodeTypeSkipped
+			}
+			newWF.Status.Nodes[node.ID] = node
+		case wfv1.NodeError, wfv1.NodeFailed, wfv1.NodeRunning:
```

Comment:
Why workflow cannot be rerun with error or failed states? Did you miss a continue under this case?

Reply:
it's a legit syntax but agree it's confusing. 
https://github.com/golang/go/wiki/Switch#noop-case
I add "continue" explicitly